include/mxnet/operator_util.h

```diff
@@ -0,0 +1,412 @@
+                              BinaryGradFunctionT1 fgrad,
+                              SimpleOpInplaceOption inplace_out_lhs_grad) = 0;
+  /*!
+   * \brief Describe the function.
+   * \param description The description of the function.
+   * \return reference to self.
+   */
+  virtual TSelf& describe(const std::string &description) = 0;
+  /*!
+   * \brief Describe the function.
+   * \param args argument information.
+   *  Add addtional arguments to the function.
+   * \return reference to self.
+   */
+  virtual TSelf& add_arguments(const std::vector<dmlc::ParamFieldInfo> &args) = 0;
+  /*! \brief virtual destructor */
+  virtual ~SimpleOpRegEntry() {}
+};
+
+/*! \brief registry for TBlob functions */
+class SimpleOpRegistry {
+ public:
+  /*!
+   * \brief Internal function to register a name function under name.
+   * \param name name of the function
+   * \return ref to the registered entry, used to set properties
+   */
+  SimpleOpRegEntry &__REGISTER_OR_FIND__(const std::string& name);
+  /*!
+   * \brief Find the entry with corresponding name.
+   * \param name name of the function
+   * \return the corresponding function, can be NULL
+   */
+  inline static const SimpleOpRegEntry *Find(const std::string &name) {
+    return Get()->fmap_.at(name);
+  }
+  /*! \return global singleton of the registry */
+  static SimpleOpRegistry* Get();
+
+ private:
+  // destructor
+  ~SimpleOpRegistry();
+  /*! \brief internal registry map */
+  std::map<std::string, SimpleOpRegEntry*> fmap_;
+};
+
+/*!
+ * \brief assign the expression to out according to request
+ * \param out the data to be assigned
+ * \param req the assignment request
+ * \param exp the expression
+ * \tparam OType output type
+ * \tparam Exp expression type
+ */
+#define ASSIGN_DISPATCH(out, req, exp)  \
+  {                                     \
+    switch (req) {                      \
+      case kNullOp:                     \
+        break;                          \
+      case kWriteTo:                    \
+      case kWriteInplace:               \
+        (out) = (exp);                  \
+        break;                          \
+      case kAddTo:                      \
+        (out) += (exp);                 \
+        break;                          \
+      default:                          \
+        LOG(FATAL) << "not reached";    \
+    }                                   \
+  }
+
+//--------------------------------------------------------------
+// The following part are API Registration of Simple Operators
+//--------------------------------------------------------------
+/*!
+ * \brief Macro to register simple operator to both imperative and symbolic API.
+ *
+ * see src/operator/elementwise_unary_op-inl.h for example
+ *
+ * \code
+ * // example of registering a sigmoid operator on GPU
+ * // MySigmoid is of type UnaryFunction,
+ * // MySigmoidGrad is of type UnaryGradFunctionT2
+ *
+ * MXNET_REGISTER_SIMPLE_OP(sigmoid, cpu)
```

Comment:
The type support is dispatched in runtime currently, see implementation of forward